src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,191 @@
+                _noExactMatchesFromNuspec.RemoveWhere(IsCompat);
+                (string noExactMatchString, string compatMatchString) =
+                    GenerateWarningString(_noExactMatchesFromFile, _noExactMatchesFromNuspec, _compatNotExactMatches);
+
+                var issues = new List<PackagingLogMessage>();
+
+                if (_noExactMatchesFromFile.Count != 0 || _noExactMatchesFromNuspec.Count != 0)
+                {
+                    issues.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat,noExactMatchString),
+                        NuGetLogCode.NU5128));
+                }
+
+                if (_compatNotExactMatches.Count != 0)
+                {
+                    issues.Add(PackagingLogMessage.CreateWarning(string.Format(SecondWarning, compatMatchString),
+                        NuGetLogCode.NU5130));
+                }
+
+                if(issues.Count != 0)
+                {
+                    return issues;
+                }
+            }
+            else
+            if((tfmsFromFiles.Count != 0 && tfmsFromNuspec.Count == 0) || (tfmsFromFiles.Count == 0 && tfmsFromNuspec.Count != 0))
+            {
+                if(tfmsFromFiles.Count != 0)
+                {
+                    _noExactMatchesFromFile.AddRange(tfmsFromFiles);
+                }
+
+                if(tfmsFromNuspec.Count != 0)
+                {
+                    _noExactMatchesFromNuspec.AddRange(tfmsFromNuspec);
+                }
+
+                (string noExactMatchString, string compatMatchString) =
+                    GenerateWarningString(_noExactMatchesFromFile, _noExactMatchesFromNuspec, _compatNotExactMatches);
+
+                var issues = new List<PackagingLogMessage>();
+                issues.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat, noExactMatchString),
+                        NuGetLogCode.NU5128));
+                return issues;
+            }
+
+            return Array.Empty<PackagingLogMessage>();
+        }
+
+        private bool IsCompat(NuGetFramework tfm)
+        {
+            return _compatNotExactMatches.Contains(tfm);
+        }
+
+        private static IEnumerable<ContentItemGroup> GetContentForPattern(ContentItemCollection collection, PatternSet pattern)
+        {
+            return collection.FindItemGroups(pattern);
+        }
+
+        private static IEnumerable<NuGetFramework> GetGroupFrameworks(IEnumerable<ContentItemGroup> groups)
+        {
+
+            return groups.Select(e => (NuGetFramework)e.Properties["tfm"]);
+        }
+
+        private static (string, string) GenerateWarningString(HashSet<NuGetFramework> noExactMatchesFromFile,
+                HashSet<NuGetFramework> noExactMatchesFromNuspec, HashSet<NuGetFramework> compatNotExactMatches)
+        {
+            string noExactMatchString = string.Empty;
+            string compatMatchString = string.Empty;
+            var beginning = "\n- Add ";
```

Comment:
These message won't get localized. 
You need to use resources for every single user facing message. 
Everything except the generated `lib/net472/a.dll` like content :)

Reply:
Well, How would I populate a resource with a variable amount of strings?

Reply:
You localize "Add". 

The resource would be add.